### PR TITLE
[vcpkg-xcode] Add new function group vcpkg_xcode_*

### PR DESCRIPTION
**Pull request by JackBoosY**
Introduce a new function group `vcpkg_xcode_*` to build Xcode project.

- [x] code
- [x] doc
- [x] test

### CODE REVIEW

**Comment by JackBoosY** [comment]
Any suggestions would be appreciated.

**Comment by dg0yt** [comment]
> Any suggestions would be appreciated.

It is only code, but no doc or test. Before commenting a particular implementation , I would like to understand the use case. Writing doc or tests often triggers immediate changes to the code.

File: ports/vcpkg-xcode/vcpkg_xcode_build.cmake [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                set(short_build_type "rel")
+                set(config "Release")
+                if (arg_ENABLE_INSTALL)
+                    set(install_param "DSTROOT=${CURRENT_PACKAGES_DIR}")
+                endif()
+            endif()
+
+            message(STATUS "Building ${TARGET_TRIPLET}-${short_build_type}")
+
+            debug_message("command:\nxcodebuild ${build_param} ${target_param} -configuration ${config} SYMROOT=${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-${short_build_type} ${install_param}")
```
Comment (JackBoosY) [inline ports/vcpkg-xcode/vcpkg_xcode_build.cmake:75]:
```suggestion
```

**Comment by Neumann-A** [comment]
> It is only code, but no doc or test. Before commenting a particular implementation , I would like to understand the use case. Writing doc or tests often triggers immediate changes to the code.

In addition to that maybe one real port using this should exist.

**Comment by JackBoosY** [comment]
> In addition to that maybe one real port using this should exist.

Yeah, choose a suitable example to add as a port is difficult for me, do you have any suggestions?

**Comment by ras0219-msft** [comment]
> choose a suitable example to add as a port is difficult for me

What port motivated creating the helper at all?

**Comment by BillyONeal** [comment]
@valeriaconde @vicroms @dan-shaw @markle11m @ras0219-msft and I discussed this today.

We are OK with this as long as a test port is added.

**Comment by JackBoosY** [comment]
> What port motivated creating the helper at all?

A private project.

**Comment by JackBoosY** [comment]
```
-- Building x64-osx-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:58 (file):
  file failed to open for reading (No such file or directory):

    /Users/vagrant/Data/buildtrees/kvocontroller/install-x64-osx-dbg-out.log
```
This error bothers me. Why it happened?

**Comment by JackBoosY** [comment]
```
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
```
@BillyONeal Requires to link to the Xcode which installed in OSX VM.

**Comment by BillyONeal** [comment]
Is there a headless way to install that?

**Comment by JackBoosY** [comment]
> Is there a headless way to install that?

https://stackoverflow.com/questions/17980759/xcode-select-active-developer-directory-error

File: ports/kvocontroller/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+    "name": "kvocontroller",
+    "version-semver": "1.2.0",
+    "description": "Key-value observing is a particularly useful technique for communicating between layers in a Model-View-Controller application.",
+    "homepage": "https://github.com/facebookarchive/KVOController",
```
Comment (dg0yt) [inline ports/kvocontroller/vcpkg.json:5]: This seems to be unmaintained since 5 years. Is there no more active consumer?

Comment (JackBoosY) [reply ports/kvocontroller/vcpkg.json:5]: But it's popular: 7354 stars.

**Comment by BillyONeal** [comment]
> > Is there a headless way to install that?
> 
> https://stackoverflow.com/questions/17980759/xcode-select-active-developer-directory-error

So.... no? There's no headless way listed there, only going through the app store...

**Comment by JackBoosY** [comment]
> > > Is there a headless way to install that?
> > 
> > 
> > https://stackoverflow.com/questions/17980759/xcode-select-active-developer-directory-error
> 
> So.... no? There's no headless way listed there, only going through the app store...

see https://developer.apple.com/downloads/index.action if it has command line mode.

File: ports/kvocontroller/vcpkg.json (hunk 2 above)
Comment (dg0yt) [reply ports/kvocontroller/vcpkg.json:5]: I have seen that. But pulling an unmaintained package into vcpkg means adding an extra maintenance burden, even more if it is popular.

File: ports/flex/vcpkg.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "flex",
```
Comment (dg0yt) [inline ports/flex/vcpkg.json:2]: Not a good name for this port:
https://repology.org/project/flex/packages

**Comment by BillyONeal** [comment]
> > So.... no? There's no headless way listed there, only going through the app store...
> see https://developer.apple.com/downloads/index.action if it has command line mode.

That goes to a login page for me. Can you supply like an actual script that will install the right parts? I don't think we have interactive access to the VMs to do a non-headless install.

**Comment by JackBoosY** [comment]
Any conclusions here? Will this PR not be accepted if I can't provide the steps to install the command line tool using pure commands?

**Comment by BillyONeal** [comment]
While not everything we have in the curated catalog is 100% tested, we tend to at least have some form of smoke test for each major feature. As long as we don't know how to get the fleet to have the prerequisites for this we can't feasibly do that. We would be happy to add this helper port but we need some way to tell we haven't broken it if you want us to be its maintainers.

Our current instructions for setting up the MacOS machines are located at https://github.com/microsoft/vcpkg/tree/master/scripts/azure-pipelines/osx#creating-a-new-vagrant-box

I have a meeting with the folks who handle this for DevDiv overall next week which might make this easier to handle.

**Comment by JackBoosY** [comment]
Needs to improve the installation process.

**Comment by JackBoosY** [comment]
Okay, sample port and test port added.

**Comment by JackBoosY** [comment]
@BillyONeal 
```
CMake Error:
  Xcode 1.5 not supported.
```
The test port requires Xcode version 3.0 or later:
https://cmake.org/cmake/help/v3.9/generator/Xcode.html

**Comment by dg0yt** [comment]
> The test port requires Xcode version 3.0 or later: https://cmake.org/cmake/help/v3.9/generator/Xcode.html

The CMake generator now requires Xcode 5.0. But IMO you are only getting reminded that XCode (GUI) isn't present in vcpkg CI.

**Comment by JackBoosY** [comment]
> > The test port requires Xcode version 3.0 or later: https://cmake.org/cmake/help/v3.9/generator/Xcode.html
> 
> The CMake generator now requires Xcode 5.0. But IMO you are only getting reminded that XCode (GUI) isn't present in vcpkg CI.

Yeah, that's the first thing which we should handle.

**Comment by BillyONeal** [comment]
Closing this draft PR due to no action in the last year. Please feel free to recreate or reopen!

**Pull request closed without merging**